### PR TITLE
Clarify intended use case for From/TryFrom/Into/TryInto

### PR DESCRIPTION
**Pull request by kangalio**
Many people seem to believe these traits should be implemented for transformations between any two types.

I decided to put the new text into std::convert itself instead of any particular trait because it applies to all four conversion traits.

### CODE REVIEW

**Comment by scottmcm** [comment]
I think the new text is reasonable, but I'd like to make sure that @rust-lang/libs-api agrees.

```text
//! # When to implement these traits
//!
//! Generally, the conversion traits in this module are intended to convert between types that
//! are semantically equivalent and differ only in representation.
//!
//! For example, [`From`]/[`TryFrom`] implementations exist for string to string, list to list,
//! number to number, or error to error conversions. However, implementing `From<i32>` for
//! `String` or `TryFrom<&str>` for `i32` would not be appropriate.
```

My instinct is that the "generally" here makes it guidance, not a rule, and thus it's fine, but I figure it's worth checking before making it "official".

**Review by Amanieu — APPROVED** [review]
LGTM

File: library/core/src/convert/mod.rs [hunk 1]
```diff
@@ -22,6 +22,15 @@
...
 //! outside the current crate.
 //!
+//! # When to implement these traits
+//!
+//! Generally, the conversion traits in this module are intended to convert between types that
+//! are semantically equivalent and differ only in representation.
+//!
+//! For example, [`From`]/[`TryFrom`] implementations exist for string to string, list to list,
+//! number to number, or error to error conversions. However, implementing `From<i32>` for
+//! `String` or `TryFrom<&str>` for `i32` would not be appropriate.
```
Comment (joshtriplett) [inline library/core/src/convert/mod.rs:32]:
```suggestion
//! number to number, or error to error conversions, as well as conversions between
//! `OwnedFd` or `OwnedHandle` and types such as `File`.
//!
//! However, don't use `From` or `TryFrom` for:
//! - Formatting, such as implementing `From<i32>` for `String`; use [`Display`] for that.
//! - Parsing, such as implementing `TryFrom<&str>` for `i32`; use [`FromStr`] for that.
```

(Needs word-wrapping.)

File: library/core/src/convert/mod.rs [hunk 2]
```diff
@@ -22,6 +22,15 @@
 //! may be necessary to implement [`Into`] or [`TryInto`] directly when converting to a type
 //! outside the current crate.
 //!
+//! # When to implement these traits
+//!
+//! Generally, the conversion traits in this module are intended to convert between types that
+//! are semantically equivalent and differ only in representation.
```
Comment (joshtriplett) [inline library/core/src/convert/mod.rs:28]: We talked about this in today's @rust-lang/libs-api meeting. We felt that the text below was helpful, but this general description didn't quite work. "semantically equivalent" doesn't feel like it matches some cases in the wild.

For instance, it's fine to have `From`/`TryFrom` implementations between subset/superset types, such as between `Ipv4Addr` and `IpAddr`.

We felt like this needs a different summary of when to use `From`. We weren't quite sure what that summary should be, just that "semantically equivalent and differ only in representation" didn't feel accurate and didn't seem sufficient to capture some cases.

**Comment by Dylan-DPC** [comment]
@kangalio any updates on this?

**Comment by Dylan-DPC** [comment]
@kangalio any updates on this?

**Comment by kangalio** [comment]
Sorry for the delay. How about just "Generally, the conversion traits in this module are intended to convert from some type to a superset type (From/Into) or to a subset type (TryFrom/TryInto)" and adjusting the examples' wording accordingly?

**Comment by workingjubilee** [comment]
Yeah, that sounds reasonable. I'm not sure about "superset type" vs. "subset type"? "A type that logically includes the other" maybe?

**Comment by Dylan-DPC** [comment]
@kangalio any updates on the recent feedback? thanks

**Comment by kangalio** [comment]
To be honest, I'm not invested into this change anymore or interested to spend the effort to get back into the mind context of this PR, so I'd regretfully go ahead and close this PR

**Pull request closed without merging**